src/logics/morphoblue/logic.flash-loan.ts

```diff
@@ -0,0 +1,124 @@
+import { MorphoFlashLoanCallback__factory, Morpho__factory } from './contracts';
+import { Service } from './service';
+import * as common from '@protocolink/common';
+import * as core from '@protocolink/core';
+import { getContractAddress, getMarkets, supportedChainIds } from './configs';
+import invariant from 'tiny-invariant';
+
+export type FlashLoanLogicTokenList = common.Token[];
+
+export type FlashLoanLogicParams = core.FlashLoanParams;
+
+export type FlashLoanLogicQuotation = core.FlashLoanQuotation;
+
+export type FlashLoanLogicFields = core.FlashLoanFields;
+
+export class FlashLoanLogic extends core.Logic implements core.LogicTokenListInterface, core.LogicBuilderInterface {
+  static id = 'flash-loan';
+  static protocolId = 'morphoblue';
+  static readonly supportedChainIds = supportedChainIds;
+
+  get callbackAddress() {
+    return getContractAddress(this.chainId, 'MorphoFlashLoanCallback');
+  }
+
+  async calcCallbackFee(loan: common.TokenAmount) {
+    const callback = MorphoFlashLoanCallback__factory.connect(this.callbackAddress, this.provider);
+    const feeRate = await callback.feeRate();
+    const callbackFee = new common.TokenAmount(loan.token).setWei(common.calcFee(loan.amountWei, feeRate.toNumber()));
+
+    return callbackFee;
+  }
+
+  async getTokenList() {
```

Comment:
Both liquidity and collateral can be used in flashloan https://docs.morpho.org/contracts/morpho-blue/reference/morpho-blue#flashloan